### PR TITLE
Export to Android using Gradle

### PR DESCRIPTION
**Pull request by amanj120**
Choosing "custom build" will allow users to create a full gradle project for their game including their assets.

### CODE REVIEW

File: platform/android/export/export.cpp [hunk 1]
```diff
@@ -2274,109 +2529,252 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+	virtual Error export_project(const Ref<EditorExportPreset> &p_preset, bool p_debug, const String &p_path, int p_flags = 0) {
+
+		ExportNotifier notifier(*this, p_preset, p_debug, p_path, p_flags);
+
+		String src_apk;
+
+		EditorProgress ep("export", "Exporting for Android", 105, true);
+
+		if (bool(p_preset->get("custom_template/use_custom_build"))) { //custom build
+			return gradle_build(p_preset, p_debug, p_path, src_apk, p_flags);
```
Comment (thanwah) [inline platform/android/export/export.cpp:2760]: Does the code below setting `src_apk` need to be moved up above this line, or should src_apk be removed from the parameter list for gradle_build?

Comment (thanwah) [reply platform/android/export/export.cpp:2760]: Looks like src_apk isn't need as a a param to gradle_build.

File: platform/android/export/export.cpp [hunk 2]
```diff
@@ -683,6 +758,22 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		return enabled_plugins;
 	}
 
+	static Error store_in_gradle_project(const String &p_path, const Vector<uint8_t> &p_data, int compression_method = Z_DEFLATED) {
+		//use dir_access.cpp and file_access.cpp to write the file
```
Comment (amanj120) [inline platform/android/export/export.cpp:762]: make comments useful or remove it

Comment (amanj120) [reply platform/android/export/export.cpp:762]: And make the method name more useful

File: platform/android/export/export.cpp [hunk 3]
```diff
@@ -732,6 +852,12 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		return OK;
 	}
 
+	static Error save_gradle_project_file(void *p_userdata, const String &p_path, const Vector<uint8_t> &p_data, int p_file, int p_total) {
```
Comment (amanj120) [inline platform/android/export/export.cpp:855]: make method name more useful

File: platform/android/export/export.cpp [hunk 4]
```diff
@@ -744,6 +870,131 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		}
+		PoolStringArray user_perms = p_preset->get("permissions/custom_permissions");
+		for (int i = 0; i < user_perms.size(); i++) {
+			String user_perm = user_perms[i].strip_edges();
+			if (!user_perm.empty()) {
+				perms->push_back(user_perm);
+			}
+		}
+		if (p_give_internet) {
+			if (perms->find("android.permission.INTERNET") == -1)
```
Comment (amanj120) [inline platform/android/export/export.cpp:893]: add brackets here?

File: platform/android/export/export.cpp [hunk 5]
```diff
@@ -744,6 +870,131 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+			int hand_tracking_index = p_preset->get("xr_features/hand_tracking"); // 0: none, 1: optional, 2: required
+			if (hand_tracking_index > 0) {
+				if (perms->find("com.oculus.permission.HAND_TRACKING") == -1) {
+					perms->push_back("com.oculus.permission.HAND_TRACKING");
+				}
+			}
+		}
+	}
+
+	Error _fix_manifest_plaintext(const Ref<EditorExportPreset> &p_preset, String &manifest_path, bool p_give_internet) {
```
Comment (amanj120) [inline platform/android/export/export.cpp:908]: add comments on all new methods

File: platform/android/export/export.cpp [hunk 6]
```diff
@@ -1339,6 +1568,7 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 			return str;
 		}
 	}
+
```
Comment (amanj120) [inline platform/android/export/export.cpp:1571]: delete newline

**Comment by akien-mga** [comment]
What does this aim to do that isn't already implemented by the Custom Build feature for Android in Godot 3.2?

For feature proposals, a detailed proposal should be opened on https://github.com/godotengine/godot-proposals to explain the use case and the proposed implementation, so that it can be discussed with the community.

CC @madmiraal

**Comment by amanj120** [comment]
[Open Issue](https://github.com/godotengine/godot-proposals/issues/1111)

**Comment by amanj120** [comment]
I am going to be closing this PR and instead submit multiple smaller, incremental PR's that will build up to do the same thing. This is to make the code reviews easier, and because having smaller, more focused PR's are generally better practice.

**Pull request closed without merging**